carma-messenger-core/cpp_message/src/cpp_message.cpp

```diff
@@ -122,314 +124,69 @@ namespace cpp_message
+            if (message->value.choice.TestMessage04.body.present == TrafficControlRequest_PR_reserved){
+                
+                output.choice = j2735_msgs::TrafficControlRequest::RESERVED;
             }
-            output.version = version;
-            // copy scale field
-            output.scale = message->value.choice.TestMessage04.body.scale;
-            // copy bounds
-            auto bounds_count = message->value.choice.TestMessage04.body.bounds.list.count;
-            for(auto i = 0; i < bounds_count; i++) {
-                j2735_msgs::ControlBounds bound;
-                // copy lat/lon
-                bound.latitude = message->value.choice.TestMessage04.body.bounds.list.array[i]->lat;
-                bound.longitude = message->value.choice.TestMessage04.body.bounds.list.array[i]->lon;
-                // copy offset array to boost vector
-                auto count = message->value.choice.TestMessage04.body.bounds.list.array[i]->offsets.list.count;
-                for(auto j = 0; j < 3; j++) {
-                    bound.offsets[j] = *message->value.choice.TestMessage04.body.bounds.list.array[i]->offsets.list.array[j];
+            else if (message->value.choice.TestMessage04.body.present == TrafficControlRequest_PR_tcrV01){
+
+                output.choice = j2735_msgs::TrafficControlRequest::TCRV01;
+                j2735_msgs::TrafficControlRequestV01 tcrV01;
+
+                // decode id
+                uint8_t id[8];
+                auto id_len = message->value.choice.TestMessage04.body.choice.tcrV01.reqid.size;
+                for(auto i = 0; i < id_len; i++)
+                {
+                    tcrV01.reqid.id[i] = message->value.choice.TestMessage04.body.choice.tcrV01.reqid.buf[i];
                 }
+
+                tcrV01.reqseq = message->value.choice.TestMessage04.body.choice.tcrV01.reqseq;
+
+                tcrV01.scale = message->value.choice.TestMessage04.body.choice.tcrV01.scale;
+
+                // copy bounds
+                auto bounds_count = message->value.choice.TestMessage04.body.choice.tcrV01.bounds.list.count;
+                for(auto i = 0; i < bounds_count; i++) {
+                j2735_msgs::TrafficControlBounds bound;
+                
                 // recover a long value from 8-bit array
                 uint64_t long_bits = 0;
-                auto bits_array_size = message->value.choice.TestMessage04.body.bounds.list.array[i]->oldest.size;
+                auto bits_array_size = message->value.choice.TestMessage04.body.choice.tcrV01.bounds.list.array[i]->oldest.size;
                 for(auto j = 0; j < bits_array_size; j++) {
-                    long_bits |= message->value.choice.TestMessage04.body.bounds.list.array[i]->oldest.buf[j];
+                    long_bits |= message->value.choice.TestMessage04.body.choice.tcrV01.bounds.list.array[i]->oldest.buf[j];
                     long_bits = long_bits << 8;
```

Comment:
Should this operation be skipped on the last turn? e.g. for 8 positions, we shift only 7 times.

Reply:
I have confirmed that adding `if (j != bits_array_size -1)` in front of `long_bits = long_bits << 8;` works.